momento-sdk/src/main/java/momento/sdk/RetryClientInterceptor.java

```diff
@@ -0,0 +1,168 @@
+import momento.sdk.retry.RetryingUnaryClientCall;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Interceptor for retrying client calls with gRPC servers. This interceptor is responsible for
+ * handling retry logic when making unary (single request, single response) gRPC calls.
+ *
+ * <p>A {@link ClientCall} is essentially an instance of a gRPC invoker. Every gRPC interceptor
+ * expects us to return such client call(s) that it will execute in order. Each call has a "start"
+ * method, which is the entry point for the call.
+ *
+ * <p>This retry client interceptor returns an instance of a {@link RetryingUnaryClientCall}, which
+ * is a client call designed to handle retrying unary (single request, single response) operations.
+ * The interceptor uses a provided {@link RetryStrategy} to determine when and how to retry failed
+ * calls.
+ *
+ * <p>When a gRPC call is intercepted, the interceptor checks whether the method is unary (client
+ * sends one message), and if so, it wraps the original {@link ClientCall} with the {@link
+ * RetryingUnaryClientCall}. This custom call is responsible for handling the retry logic.
+ *
+ * <p>When the gRPC call is closed, the {@code onClose} method is called, which is the point where
+ * we can safely check the status of the initial request that was made and determine if we want to
+ * retry or not. It starts the retrying process by scheduling a new call attempt with a delay
+ * provided by the {@link RetryStrategy}.
+ *
+ * <p>If the retry attempts are exhausted or if the provided delay is not present (indicating that
+ * we should not retry anymore), the interceptor propagates the final result to the original
+ * listener, effectively completing the call with the last status received.
+ *
+ * <p>Note that the interceptor only supports unary operations for retrying.
+ *
+ * @see RetryStrategy
+ * @see RetryEligibilityStrategy
+ * @param <ReqT> The type of the request message.
+ * @param <RespT> The type of the response message.
+ */
+final class RetryClientInterceptor implements ClientInterceptor {
+
+  private final RetryStrategy retryStrategy;
+  private final ScheduledExecutorService scheduler;
+  private final ExecutorService executor;
+  private final Logger logger = LoggerFactory.getLogger(RetryClientInterceptor.class);
+
+  public RetryClientInterceptor(
+      final RetryStrategy retryStrategy,
+      final ScheduledExecutorService scheduler,
+      final ExecutorService executor) {
+    this.retryStrategy = retryStrategy;
+    this.scheduler = scheduler;
+    this.executor = executor;
+  }
+
+  @Override
+  public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+      final MethodDescriptor<ReqT, RespT> method,
+      final CallOptions callOptions,
+      final Channel channel) {
+    // currently the SDK only supports unary operations which we want to retry on
+    if (!method.getType().clientSendsOneMessage()) {
+      return channel.newCall(method, callOptions);
```

Comment:
Will this be intercepted again?

Reply:
This is the client call which doesn't return our interceptor if it's a non-unary operation. The unary operations are the ones that provide the RetryingClientCall and those will be intercepted. I did a quick test on my end where I added ```NOT_FOUND``` as a retryable status and put my cache name as a random string. I enabled the log and saw 3 attempts being made.